File: Offer 29. 顺时针打印矩阵.java
class Solution {
    public int[] spiralOrder(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return new int[0];
        }

        int m = matrix.length, n = matrix[0].length;
        int[] res = new int[m * n];
        int index = 0;
        // 定义四个边界
        int left = 0, right = n - 1, top = 0, bottom = m - 1;
        while (left <= right && top <= bottom) {
            for (int c = left; c <= right; c++) {
                // 上边
                res[index++] = matrix[top][c];
            }

            for (int r = top + 1; r <= bottom; r++) {
                // 右边
                res[index++] = matrix[r][right];
            }

            if (left != right && top != bottom) {
                for (int c = right - 1; c >= left; c--) {
                    // 下边
                    res[index++] = matrix[bottom][c];
                }

                for (int r = bottom - 1; r > left; r--) {
                    // 左边
                    res[index++] = matrix[r][left];
                }
            }

            // 进入下一层
            top++;
            bottom--;
            left++;
            right--;
        }

        return res;
    }
}

File: Offer 29. 顺时针打印矩阵.py
#模拟边界
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix or len(matrix) == 0:
            return []
        
        m, n = len(matrix), len(matrix[0])
        left, right, top, bottom = 0, n - 1, 0, m - 1

        ans = []
        while left <= right and top <= bottom:
            for c in range(left, right + 1):
                ans.append(matrix[top][c])
            for r in range(top + 1, bottom + 1):
                ans.append(matrix[r][right])

            if top < bottom and left < right:
                for c in range(right - 1, left - 1, -1):
                    ans.append(matrix[bottom][c])

                for r in range(bottom - 1, top, -1):
                    ans.append(matrix[r][left])

            left += 1
            right -= 1
            top += 1
            bottom -= 1

        return ans

# python简易写法
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        while matrix:
            ans += matrix.pop(0)
            matrix = list(zip(*matrix))[::-1]
        return ans